### PR TITLE
Font options are not toggleable any longer

### LINKED ISSUE #584: Fonts setting should not be toggling options
## Are you reporting a feature or a bug?

Feature request

## [Check if the issue is already reported](http://docs.ckeditor.com/#!/guide/dev_issues_readme-section-avoid-duplicates)

http://dev.ckeditor.com/ticket/16865

## Provide detailed reproduction steps (if any)

1. Change the font size for the first time it works correctly.
2. But if select the same font size again, it will revert back to the original font size. The same for font style.
3. This behavior doesn't conform to other main editors, and could bring problems. 

### Expected result

Please prevent revert font back to original font if select same font again.

### Actual result

Resetting the font to original font if select font again.

## Additional Information

To provide a possibility to remove the font all the way. It was specified in [the comment below](https://github.com/ckeditor/ckeditor-dev/issues/584#issuecomment-313076114).

### PR DESCRIPTION
**Pull request by msamsel**
## What is the purpose of this pull request?
New feature

## Does your PR contain necessary tests?
yes

### This PR contains
- [x] Unit tests
- [x] Manual tests

## What changes did you make?
- Disable possibility to toggle option on drop down list of: **font size**, **font family name**, **paragraph format**
- Add `(Default)` option to **font size** and **font family** name, which remove formating from selection
- Add new language entry with `Default` to `lang.common` file
- Fix one unit test which click twice into this same menu option and start to fail after change

close #584

### CODE REVIEW

File: plugins/font/plugin.js [hunk 1]
```diff
@@ -87,11 +87,17 @@
...
 			init: function() {
+				var name,
+					defaultValue = 'cke-default',
+					defaultText = '(' + editor.lang.common.optionDefault + ')';
+
 				this.startGroup( lang.panelTitle );
 
-				for ( var i = 0; i < names.length; i++ ) {
-					var name = names[ i ];
+				// Add `(Default)` item as first element on the drop down list.
```
Comment (f1ames) [inline plugins/font/plugin.js]: > // Add `(Default)` item as **a** first element on the drop down list.

File: plugins/font/plugin.js [hunk 2]
```diff
@@ -102,27 +108,35 @@
...
+					style = styles[ value ],
+					previousStyle,
+					range,
+					path,
+					matching,
+					startBoundary,
+					endBoundary,
+					node,
+					bm,
+					defaultValue = 'cke-default';
```
Comment (f1ames) [inline plugins/font/plugin.js]: Same default value is already defined in an `init` method, maybe it could be extracted somewhere to be more "global"? Especially that it is used also in other places (tests).

File: plugins/format/plugin.js [hunk 3]
```diff
@@ -64,10 +64,10 @@ CKEDITOR.plugins.add( 'format', {
 				editor.focus();
 				editor.fire( 'saveSnapshot' );
 
-				var style = styles[ value ],
-					elementPath = editor.elementPath();
+				var style = styles[ value ];
 
-				editor[ style.checkActive( elementPath, editor ) ? 'removeStyle' : 'applyStyle' ]( style );
+				// Always aply style, do not allow on toggle it by clicking on list item (#584).
```
Comment (f1ames) [inline plugins/format/plugin.js]: > // Always **apply** style, do not allow on toggle it by clicking on list item (#584).

File: plugins/format/plugin.js [hunk 4]
```diff
@@ -64,10 +64,10 @@ CKEDITOR.plugins.add( 'format', {
...
+				editor.applyStyle( style );
```
Comment (f1ames) [inline plugins/format/plugin.js]: By default all format tags are block elements so reapplying them to partial selection doesn't do anything. However, if you change the config like:

```
config.format_p = { element: 'span', attributes: { 'class': 'normalPara' } };
```

You can apply it to partial selection despite the fact it was already applied (so it creates nested tags). This is an edge case IMHO but still we should handle this. You could just check if style is active (as before `style.checkActive( elementPath, editor )`) and apply it only if it is not already there.

File: tests/plugins/font/font.js [hunk 5]
```diff
@@ -170,6 +172,43 @@
 				'<p>x<span style="font-size:12px"><em>foo</em></span><em><span style="font-size:24px">bar</span></em>x@</p>' );
 		},
 
+		// #584
```
Comment (f1ames) [inline tests/plugins/font/font.js]: Is there any test checking if second click (reapplying) of the same font/size doesn't change the content?

File: tests/plugins/font/manual/preservestyle.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
+@bender-tags: 4.8.0, feature, 584
+@bender-ui: collapsed
+@bender-ckeditor-plugins: wysiwygarea, toolbar, font, elementspath, sourcearea
+
+1. Select text
```
Comment (f1ames) [inline tests/plugins/font/manual/preservestyle.md]: Each list item should end with `.`. `Select some text` sounds a little better.

File: tests/plugins/font/font.js (hunk 5 above)
Comment (f1ames) [inline tests/plugins/font/font.js]: I see there is on in a `format` plugin. Still those are separate plugin with different code so it should be tested for both.

**Comment by msamsel** [comment]
I hope it is fixed now :)

File: plugins/format/plugin.js [hunk 7]
```diff
@@ -67,7 +67,10 @@ CKEDITOR.plugins.add( 'format', {
 				var style = styles[ value ],
 					elementPath = editor.elementPath();
 
-				editor[ style.checkActive( elementPath, editor ) ? 'removeStyle' : 'applyStyle' ]( style );
+				// Always apply style, do not allow on toggle it by clicking on list item (#584).
```
Comment (f1ames) [inline plugins/format/plugin.js]: > Always apply style, do not allow **to** toggle it by clicking on **corresponding** list item (#584).

?

File: tests/plugins/font/font.js [hunk 8]
```diff
@@ -35,8 +48,10 @@
 
 				this.wait( function() {
 					// Click again to exit the style.
+					// Since 4.8.0 2nd click into this same menu item do not unselect it.
```
Comment (f1ames) [inline tests/plugins/font/font.js]: > Since 4.8.0, 2nd click **on the** same menu item does not unselect it.

File: tests/plugins/font/font.js [hunk 9]
```diff
@@ -35,8 +48,10 @@
...
+					// It is required to click into '(Default)' option to reset style (#584).
```
Comment (f1ames) [inline tests/plugins/font/font.js]: > It is required to click **on the** '(Default)' option to reset style (#584).

File: tests/plugins/format/plugin.js [hunk 10]
```diff
@@ -25,5 +25,24 @@ bender.test( {
...
+			combo = editor.ui.get( name );
+
+		bot.setHtmlWithSelection( '<h1>f^oo</h1>' );
+
+		// Apply the same style format to element already styled.
+		assert.areSame( CKEDITOR.TRISTATE_OFF, combo.getState(), 'check state OFF' );
+		bot.combo( name, function( combo ) {
+			assert.areSame( CKEDITOR.TRISTATE_ON, combo.getState(), 'check state ON when opened' );
+			combo.onClick( 'h1' );
+			assert.areSame( '<h1>f^oo</h1>', bot.htmlWithSelection(), 'applied h1 block style' );
```
Comment (f1ames) [inline tests/plugins/format/plugin.js]: > applied h1 block style

`h1` was not applied here as it was already there. The test (and assert) check if `h1` is not removed when applied second time, so maybe _preserved h1 block style_ or something similar?

**Comment by msamsel** [comment]
It seems that sometimes IE8 on some system can have strange behaviour. I removed checking if the format panel is opened and left only checking styles if it's properly applied.
